pex/pep_376.py

```diff
@@ -84,43 +129,267 @@ class InstalledFile(object):
+                    # We only try to link regular files since linking a symlink on Linux can produce
+                    # another symlink, which leaves open the possibility the src target could later
+                    # go missing leaving the dst dangling.
+                    if link and not os.path.islink(src):
+                        try:
+                            os.link(src, dst)
+                            continue
+                        except OSError as e:
+                            if e.errno != errno.EXDEV:
+                                raise e
+                            link = False
+                    shutil.copy(src, dst)
+                except (IOError, OSError) as e:
+                    if e.errno != errno.EEXIST:
+                        raise e
+                finally:
+                    yield src, dst
+
+    def _reinstall_site_packages(self, site_packages_dir, symlink=False):
+        link = True
+        for root, dirs, files in os.walk(self.prefix_dir, topdown=True, followlinks=True):
+            if root == self.prefix_dir:
+                dirs[:] = [d for d in filter_pyc_dirs(dirs) if d != self.stash_dir]
+                files[:] = [f for f in filter_pyc_files(files) if f != self._LAYOUT_JSON_FILENAME]
+
+            traverse = set(dirs)
+            for path, is_dir in itertools.chain(
+                zip(dirs, itertools.repeat(True)), zip(files, itertools.repeat(False))
+            ):
+                src_entry = os.path.join(root, path)
+                dst_entry = os.path.join(
+                    site_packages_dir, os.path.relpath(src_entry, self.prefix_dir)
+                )
+                try:
+                    if symlink and not (
+                        src_entry.endswith(".dist-info") and os.path.isdir(src_entry)
+                    ):
+                        dst_parent = os.path.dirname(dst_entry)
+                        safe_mkdir(dst_parent)
+                        rel_src = os.path.relpath(src_entry, dst_parent)
+                        os.symlink(rel_src, dst_entry)
+                        traverse.discard(path)
+                    elif is_dir:
+                        safe_mkdir(dst_entry)
+                    else:
+                        # We only try to link regular files since linking a symlink on Linux can
+                        # produce another symlink, which leaves open the possibility the src_entry
+                        # target could later go missing leaving the dst_entry dangling.
+                        if link and not os.path.islink(src_entry):
+                            try:
+                                os.link(src_entry, dst_entry)
+                                continue
+                            except OSError as e:
+                                if e.errno != errno.EXDEV:
+                                    raise e
+                                link = False
+                        shutil.copy(src_entry, dst_entry)
```

Comment:
Perhaps use a helper function for this highlighted code block since it's duplicated with the above. Ack on the `link` variable being in an outer scope; could maybe have the helper return a bool, then caller updates `link` based on that.

(FYI these two functions have been the hardest for me to understand - symlinks in general trip me up still.)

Reply:
I can't understand any of this at all! I run this stuff and poke around in the ~/.pex/ root to view the structure and figure out what's wrong, fix it up. It's amazingly complex to browse, but very instructive. Especially a --venv mode ~/.pex/venvs/ ... structure.

Reply:
I'll see about the factoring. It bothers me too, but its also super-titchy and I spend hours fixing this up after tweaks. Luckily tests bleed volumes when any of this is wrong.

Reply:
I'm going to defer this to a follow up.